### PR TITLE
Wip fusable fanin fanout drewhk

### PR DESCRIPTION
**Pull request by drewhk**
**PREVIEW, Not yet for merge**

Preview of fusable fan-in and fan-out stages. Still missing:
- ZipWith 2..N (there is a Zip)
- Unzip and UnzipWith 2..N
- FlexiMerge and FlexiRoute

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 1]
```diff
@@ -39,15 +37,71 @@ object Merge {
...
+
+    private def dequeueAndDispatch(): Unit = {
+      val in = pendingQueue(pendingHead % inputPorts)
+      pendingHead += 1
+      push(out, grab(in))
+      if (upstreamsClosed && !pending) completeStage()
+      else tryPull(in)
+    }
+
+    private def tryPull(in: Inlet[T]): Unit = if (!isClosed(in)) pull(in)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: looks like `tryPull` could be pulled up into `GraphStage`?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 2]
```diff
@@ -126,15 +255,61 @@ object Broadcast {
...
+            if (!isClosed(o)) {
+              push(o, elem)
+              pending(i) = true
+            }
+        }
+      }
+    })
+
+    private def tryPull(): Unit =
+      if (pendingCount == 0 && !hasBeenPulled(in)) pull(in)
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: ok, I see, `tryPull` may mean different things in different stages.

**Comment by ktoso** [comment]
Looks very good so far. I did not read very many GraphStages before, have to say the API grows on me a lot :)

**Comment by ktoso** [comment]
Reviewed the next batch of stages, LGTM

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterSpec.scala [hunk 3]
```diff
@@ -14,10 +15,10 @@ class GraphInterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
     // Reusable components
     val identity = new Identity[Int]
     val detacher = new Detacher[Int]
-    val zip = new Zip[Int, String]
-    val bcast = new Broadcast[Int](2)
-    val merge = new Merge[Int](2)
-    val balance = new Balance[Int](2)
+    val zip = Zip[Int, String]
+    val bcast = Broadcast[Int](2)
```
Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterSpec.scala]: random thought: we should probably cache these instances (up to, say, 10)

Comment (drewhk) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/GraphInterpreterSpec.scala]: it occured to me, too.

File: akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala [hunk 4]
```diff
@@ -259,6 +259,14 @@ object Concat {
    * in the `FlowGraph`. This method creates a new instance every time it
    * is called and those instances are not `equal`.
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala]: Is this comment not completely wrong by now?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 5]
```diff
@@ -39,15 +34,71 @@ object Merge {
  *
  * '''Cancels when''' downstream cancels
  */
-class Merge[T] private (inputPorts: Int,
-                        override val shape: UniformFanInShape[T, T],
-                        private[stream] override val module: StreamLayout.Module)
-  extends Graph[UniformFanInShape[T, T], Unit] {
+class Merge[T](private val inputPorts: Int) extends GraphStage[UniformFanInShape[T, T]] {
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: why is the count private while the Vector is public?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 6]
```diff
@@ -39,15 +34,71 @@ object Merge {
...
+    private def tryPull(in: Inlet[T]): Unit = if (!isClosed(in)) pull(in)
+
+    in.foreach { i ⇒
+      setHandler(i, new InHandler {
+        override def onPush(): Unit = {
+          if (isAvailable(out)) {
+            if (!pending) {
+              push(out, grab(i))
+              tryPull(i)
+            } else {
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: This branch should not ever be taken, should it?

File: akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala (hunk 4 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Graph.scala]: yes, it seems to be from an old age...

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 5 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: no need for it to be private at all

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala (hunk 6 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: Hmm, I guess not. that simplifies things.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: also for Balance

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: You are right, removed these everywhere.

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/UnzipWithApply.scala.template [hunk 7]
```diff
@@ -44,14 +42,49 @@ trait UnzipWithApply {
...
-                         private[stream] override val module: StreamLayout.Module)
-  extends Graph[FanOutShape1[In, [#A1#]], Unit] {
+class UnzipWith1[In, [#A1#]](unzipper: In ⇒ ([#A1#])) extends GraphStage[FanOutShape1[In, [#A1#]]] {
+  override val shape = new FanOutShape1[In, [#A1#]]("UnzipWith1")
+  def in = shape.in
+
+  [#def out0: Outlet[A1] = shape.out0#
+  ]
+
+  override def createLogic = new GraphStageLogic {
```
Comment (patriknw) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/UnzipWithApply.scala.template]: perhaps use an explicit return type here?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 8]
```diff
@@ -39,15 +34,68 @@ object Merge {
  *
  * '''Cancels when''' downstream cancels
  */
-class Merge[T] private (inputPorts: Int,
-                        override val shape: UniformFanInShape[T, T],
-                        private[stream] override val module: StreamLayout.Module)
-  extends Graph[UniformFanInShape[T, T], Unit] {
+class Merge[T](val inputPorts: Int) extends GraphStage[UniformFanInShape[T, T]] {
+  val in = Vector.tabulate(inputPorts)(i ⇒ Inlet[T]("Merge.in" + i))
+  val out = Outlet[T]("Merge.out")
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: explicit type?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 9]
```diff
@@ -126,15 +246,61 @@ object Broadcast {
...
+  override def createLogic: GraphStageLogic = new GraphStageLogic {
+    private var pendingCount = outputPorts
+    private val pending = Array.fill[Boolean](outputPorts)(true)
+    private var downstreamsRunning = outputPorts
+
+    setHandler(in, new InHandler {
+      override def onPush(): Unit = {
+        pendingCount = downstreamsRunning
+        val elem = grab(in)
+        out.iterator.zipWithIndex.foreach {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: while?

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 10]
```diff
@@ -126,15 +246,61 @@ object Broadcast {
...
+              pending(i) = true
+            }
+        }
+      }
+    })
+
+    private def tryPull(): Unit =
+      if (pendingCount == 0 && !hasBeenPulled(in)) pull(in)
+
+    out.iterator.zipWithIndex.foreach {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: while?

**Comment by patriknw** [comment]
LGTM, aside from some nitpicks

**Comment by drewhk** [comment]
Any objections to merge? @rkuhn @ktoso?

**Comment by rkuhn** [comment]
LGTM

**Comment by ktoso** [comment]
Go ahead (as you already did :-))

**Pull request merged**